### PR TITLE
Add extra spaces in julia_code() printing.

### LINKED ISSUE #23667: julia_code generates invalid Julia code
The [documentation for `julia_code`](https://docs.sympy.org/latest/modules/printing.html?highlight=julia_code#sympy.printing.julia.julia_code) shows a pretty simple example of generating Julia code:
```python
>>> julia_code(x**2*y*A**3)
'(x.^2.*y)*A^3'
```
Unfortunately, that snippet will produce an error if you actually try to use it in Julia:
```
syntax: invalid syntax "2.*"; add space(s) to clarify
```
The problem is an unfortunate ambiguity in Julia's syntax (which will not be changed).  The reason is that `2.` is one thing (the floating-point number 2), and `.*` is another thing (broadcasting — aka Hadamard — multiplication), and julia can't tell which one you mean, so it results in an error.

In this case, the correct code needs at least one extra space, between the `2` and the `.*`:
```julia
(x.^2 .*y)*A^3
```
I don't know if it's worth looking for just this case of a literal integer followed by a `.`, but a simple and robust alternative would be to emit code that is *closer* to most code styles anyway, and just put spaces around binary operators by default, like this:
```julia
(x .^ 2 .* y) * A ^ 3
```
Of course, most code styles also make an exception for `^`, and suggest that no spaces need to be used — but leave the door open for using them.  But here, it's much safer to include them, in case you have an expression like `2.^x`, which would result in a similar error.

### PR DESCRIPTION
**Pull request by cocolato**
#### References to other Issues or PRs

Fix #23667. 

#### Brief description of what is fixed or changed
Add extra spaces around binary operators to the result of `julia_code()` by default.
The case in #23667, now it will print `(x .^ 2 .* y) * A ^ 3` .
#### Release Notes

* printing
  * Add extra spaces around binary operators to the result of `julia_code()` by default.

### CODE REVIEW

File: sympy/printing/julia.py [hunk 1]
```diff
@@ -405,7 +404,15 @@ def _print_HadamardPower(self, expr):
             self.parenthesize(expr.exp, PREC)
             ])
 
-    # Note: as of 2015, Julia doesn't have spherical Bessel functions
+    def _print_Rational(self, expr):
+        if expr.q == 1:
+            return str(expr.p)
+        else:
+            if self._settings.get("sympy_integers", False):
+                return "S(%s) / %s" % (expr.p, expr.q)
```
Comment (oscargus) [inline sympy/printing/julia.py:412]: I am not sure where this comes from? `S` is pretty SymPy-specific and I doubt that Julia has that, but I can be wrong of course...

File: sympy/printing/julia.py [hunk 2]
```diff
@@ -177,18 +177,17 @@ def multjoin(a, a_str):
             r = a_str[0]
             for i in range(1, len(a)):
                 mulsym = '*' if a[i-1].is_number else '.*'
-                r = r + mulsym + a_str[i]
+                r = "%s %s %s" %(r, mulsym, a_str[i])
```
Comment (oscargus) [inline sympy/printing/julia.py:180]: Please add a space after `%` here and below.

File: sympy/printing/julia.py (hunk 1 above)
Comment (cocolato) [reply sympy/printing/julia.py:412]: Thanks for reminding, I copied it from `StrPrinter`, this should be deleted in `JuliaCodePrinter`.

**Comment by cocolato** [comment]
Thanks! And I wonder if the current commit is as expected?

File: sympy/printing/julia.py [hunk 3]
```diff
@@ -177,18 +177,17 @@ def multjoin(a, a_str):
...
             return sign + multjoin(a, a_str)
         elif len(b) == 1:
             divsym = '/' if b[0].is_number else './'
-            return sign + multjoin(a, a_str) + divsym + b_str[0]
+            return "%s %s %s" % (sign+multjoin(a, a_str), divsym, b_str[0])
         else:
             divsym = '/' if all(bi.is_number for bi in b) else './'
-            return (sign + multjoin(a, a_str) +
-                    divsym + "(%s)" % multjoin(b, b_str))
+            return "%s %s %s" % (sign + multjoin(a, a_str), divsym, "(%s)" % multjoin(b, b_str))
```
Comment (oscargus) [inline sympy/printing/julia.py:190]:
```suggestion
            return "%s %s (%s)" % (sign + multjoin(a, a_str), divsym, "(%s)" % multjoin(b, b_str))
```
I think.

File: sympy/printing/julia.py [hunk 4]
```diff
@@ -207,18 +206,18 @@ def _print_Pow(self, expr):
         if expr.is_commutative:
             if expr.exp == -S.Half:
                 sym = '/' if expr.base.is_number else './'
-                return "1" + sym + "sqrt(%s)" % self._print(expr.base)
+                return "%s %s %s" % ("1", sym, "sqrt(%s)" % self._print(expr.base))
```
Comment (oscargus) [inline sympy/printing/julia.py:209]:
```suggestion
                return "1 %s sqrt(%s)" % (sym, self._print(expr.base))
```

File: sympy/printing/julia.py [hunk 5]
```diff
@@ -207,18 +206,18 @@ def _print_Pow(self, expr):
...
             if expr.exp == -S.One:
                 sym = '/' if expr.base.is_number else './'
-                return "1" + sym + "%s" % self.parenthesize(expr.base, PREC)
+                return  "%s %s %s" % ("1", sym, "%s" % self.parenthesize(expr.base, PREC))
```
Comment (oscargus) [inline sympy/printing/julia.py:212]:
```suggestion
                return  "1 %s %s" % (sym, self.parenthesize(expr.base, PREC))
```

**Comment by moble** [comment]
@oscargus Sure thing.  In the distant past, I've even made commits to `sympy.printing`, so I'm right at the intersection!  :)

**Review by moble — COMMENTED** [review]
This PR does what it set out to do, and definitely improves the quality of the emitted Julia code — and its prettiness :).  I've actually run the examples in the documentation, and more of them work now.  So if nothing else, I'd give the thumbs up.

Now, I don't mean to make perfect the enemy of good, but I have a couple other suggestions that might improve things even more.

As noted, I do think Rationals should get that extra slash.  However, in reviewing the new docs, I also found another source of problems, and I'm not sure how easy it would be to fix them.  Specifically, for expressions like `sin(pi * x .* y)`, this will also fail if `x` and/or `y` is an array because `sin` also needs to be broadcast: `sin.(pi * x .* y)`.  For consistency, all functions would need to look at the contents of their arguments and decide whether or not to broadcast...

This made me realize that there's a completely different and simpler approach that may be preferable.  A relatively recent improvement to julia is the introduction of [the `@.` macro](https://docs.julialang.org/en/v1/base/arrays/#Base.Broadcast.@__dot__), which just goes through and applies `.` everywhere it makes sense.  So rather than sympy emitting expressions that include dots for some — but not all — operations, maybe you could just tell the user (or add an option to the printer) to prefix the expression with `@.`.  In fact, a *very* ugly hack that I've just found to work well is to include the `@.` in the `assign_to` argument, as in
```python
>>> julia_code(sin(pi*x*y), assign_to="@. s")
'@. s = sin(pi * x .* y)'
```
(Here, `s` has to already exist for that to work.)  

My point is that that's so much easier for the user to do that it might be better to just eliminate the dots altogether.  I'd still keep the extra spaces for most of the operators, just because it's more in line with the standard styles, but maybe get rid of them for `^`.

The only problem with this approach is that it would screw up the nice things that currently happen with matrices: `@. A^3` would do element-wise cubing, rather than matrix-wise.  But I think there are unavoidable tradeoffs here, and the user always has to be at least mildly aware of what they want.  And in this case, I think the desire for matrix-wise powers is much less common than for users to desire broadcasting everywhere.

**Comment by asmeurer** [comment]
> This made me realize that there's a completely different and simpler approach that may be preferable. A relatively recent improvement to julia is the introduction of [the @. macro](https://docs.julialang.org/en/v1/base/arrays/#Base.Broadcast.@__dot__), which just goes through and applies . everywhere it makes sense. So rather than sympy emitting expressions that include dots for some — but not all — operations, maybe you could just tell the user (or add an option to the printer) to prefix the expression with @.. In fact, a very ugly hack that I've just found to work well is to include the @. in the assign_to argument, as in

It really depends on the standard style is for Julia, which I'll have to defer to you on. 

Based on your description, it sounds like the best course of action would be to have a flag in the printer to disable dots, but it sounds like the default should be to include them. IMO, this doesn't seem like a very good design decision on Julia's part (but I don't really use Julia, so maybe I'm missing something). 

Either way, let's discuss this in a separate issue/PR, since it's unrelated to the original point of this PR (avoiding the `2.*x` invalid syntax).

> My point is that that's so much easier for the user to do that it might be better to just eliminate the dots altogether. I'd still keep the extra spaces for most of the operators, just because it's more in line with the standard styles, but maybe get rid of them for ^.

How is it easier or harder to use dots? The code is automatically generated.

My expectation would be that the `@.` is harder because there could be cases where the generated code is not being used as the rhs of an assignment. And even if it is, removing the docs would make this implicit and nonobvious in the printer. That's why to me having a flag with dots by default seems the best.

> The only problem with this approach is that it would screw up the nice things that currently happen with matrices: @. A^3 would do element-wise cubing, rather than matrix-wise. But I think there are unavoidable tradeoffs here, and the user always has to be at least mildly aware of what they want. And in this case, I think the desire for matrix-wise powers is much less common than for users to desire broadcasting everywhere.

That's the problem with making it implicit in the printer. SymPy already knows if the `^` represents a matrix power or a regular power, because one would be `Pow` and one would be `MatPow`.

**Comment by asmeurer** [comment]
> My expectation would be that the @. is harder because there could be cases where the generated code is not being used as the rhs of an assignment. And even if it is, removing the docs would make this implicit and nonobvious in the printer. That's why to me having a flag with dots by default seems the best.

Actually it seems like `@.` can be used in any expression, not just an assignment:

```julia
julia> x = [1, 2, 3]
3-element Vector{Int64}:
 1
 2
 3

julia> x .+ (@. (x + 1))
3-element Vector{Int64}:
 3
 5
 7
```

So maybe we can use it, as long as there are no matrix operations in the expressions.

File: sympy/printing/julia.py [hunk 6]
```diff
@@ -530,19 +552,19 @@ def julia_code(expr, assign_to=None, **settings):
...
     >>> x = symbols('x')
     >>> julia_code(sin(x).series(x).removeO())
-    'x.^5/120 - x.^3/6 + x'
+    'x .^ 5 / 120 - x .^ 3 / 6 + x'
 
     >>> from sympy import Rational, ceiling
     >>> x, y, tau = symbols("x, y, tau")
     >>> julia_code((2*tau)**Rational(7, 2))
-    '8*sqrt(2)*tau.^(7/2)'
+    '8 * sqrt(2) * tau .^ (7 / 2)'
```
Comment (oscargus) [inline sympy/printing/julia.py:560]:
```suggestion
    '8 * sqrt(2) * tau .^ (7 // 2)'
```

**Comment by oscargus** [comment]
Not sure if/why this breaks the Aesara-printing on pypy and 3.11...

**Comment by cocolato** [comment]
> Not sure if/why this breaks the Aesara-printing on pypy and 3.11...

Looks like it was fixed in the https://github.com/sympy/sympy/pull/23734.

**Comment by moble** [comment]
> Actually it seems like @. can be used in any expression, not just an assignment:

Right.  Sorry if my mention of that ugly hack was confusing; that's just one example of how it can be used.  But also note that `@.` can just be put in front of the entire left-hand side.  That is, you wouldn't actually use `x .+ (@. (x + 1))`, you'd use `@. x + (x + 1)`, with no other dots appearing in the expression.  Maybe a better example would be what is currently output as the incorrect `2.^x.+sin(x)`.  This fails because `2.^x` is ambiguous and because `sin` doesn't know how to operate on an array.  The first would be fixed by this PR, but the second would not yet.  The correct (and more Julian) approach would be the simpler `@. 2^x + sin(x)`.  The downside of this is that this approach would require special handling if the same expression needs both broadcasting and matrix operations.

So basically, I see three reasonable approaches.  The first would be to ignore `@.`, continue with this PR as is, but extend it so that functions like `sin(x)` become `sin.(x)`.  Matrix operations would still appear undotted — like `A ^ 3`, or `A * B`.

The second approach would be to use `@.` by default (or maybe as an option), but then you would have to escape `MatPow` and `MatMul`, so that `MatPow(A, 3)` becomes `$^(A, 3)` and `MatMul(A, B)` becomes `$*(A, B)`.  That's pretty ugly, but so are lots of dots scattered throughout an expression.

The third approach would be to just remove dots entirely, and remind the user in the documentation that `@.` exists, and let them handle it.  They have to write code around sympy's output anyway, so this isn't much more to ask.

Personally, I would favor the second approach, just because I would have *lots* of dots appearing throughout my expressions, but I almost never use matrix operations.  Maybe my view is skewed.

**Comment by asmeurer** [comment]
> That is, you wouldn't actually use x .+ (@. (x + 1)), you'd use @. x + (x + 1),

The point is if someone generates an expression with `julia_code`, then that expression can be used anywhere, even as a subexpression of a larger expression. 

It also means that we can use it recursively if parts of an expression use non-dotted operations like matmul.

**Comment by asmeurer** [comment]
I don't want to block this PR on this. The changes in this PR are already pretty self contained and seem fine to me. The dot stuff is already that way in master, so there's no reason it should block merging this. 

The test failure is unrelated. I just merged https://github.com/sympy/sympy/pull/23734 so it should start passing now. If there are no objections from anyone, we can merge it, and work on the dot stuff in a new PR.

**Comment by asmeurer** [comment]
> The second approach would be to use @. by default (or maybe as an option), but then you would have to escape MatPow and MatMul, so that MatPow(A, 3) becomes $^(A, 3) and MatMul(A, B) becomes $*(A, B). That's pretty ugly, but so are lots of dots scattered throughout an expression.

I didn't realize you could have an escaped version of a non-dotted operator too. Probably the best choice here is to have a flag that lets you pick how things are printed. Maybe the default could be something that tries to guess from an expression what to use based on which kinds of operators appear more, for instance, one way for matrix expressions and one way for Expr expressions (question: does MatAdd need to use `.+` or can it use `+`?). IMO printing the dots (or @.) by default is a good idea so that the resulting expression always automatically vectorizes, but there should be a way to disable them if the user knows the variables are scalars and wants to clean up the output a bit.

**Comment by oscargus** [comment]
> But it looks like this case isn't supported in sympy???

Correct (if I understand you correctly). SymPy only support "matrix-correct" operations.

**Comment by oscargus** [comment]
Thanks @cocolato !

**Pull request merged**